runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaCacheClientProduceFactory.java

```diff
@@ -681,6 +688,22 @@ private void onClientInitialData(
                     break init;
                 }
 
+                if (type != null && type.key != null)
+                {
+                    Validator keyValidator = type.key;
+                    if (keyValidator != null && key != null)
+                    {
+                        OctetsFW entryKey = key.value();
+                        if (entryKey != null ||
+                            !keyValidator.write(entryKey.value(), entryKey.offset(), entryKey.sizeof()))
+                        {
+                            System.out.println("Key Validation failed");
+                            error = ERROR_INVALID_RECORD;
+                            break init;
+                        }
+                    }
+                }
+
```

Comment:
This should be in `partition.writeProduceEntryStart` or else something similar to `partition.validProduceEntry` instead of spreading the validation both inside and outside the cache.